### PR TITLE
prov/gni: fixes for vc setup

### LINKED ISSUE #661: osu_alltoallv using craypich/ofi netmod hangs periodically
@kim-mcmahon reported that she still sees occasional hangs for osu_alltoallv when linked against craypich/ofi netmod.

Per the readme she supplied

```

1. SET this two env variables

export MPICH_NEMESIS_NETMOD=ofi
export MPIR_CVAR_OFI_USE_PROVIDER=gni


2. Put the following in a loop and run on tiger

srun -n 64 --ntasks-per-node=8 --exclusive ./osu_alltoallv
```

this is on tiger.

### PR DESCRIPTION
**Pull request by hppritcha**
fixes ofi-cray/libfabric-cray#661

Signed-off-by: Howard Pritchard howardp@lanl.gov
@sungeunchoi

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
well there's something messed up with edison/cori and gnu and craypich modules.  the jekins failure can be ignored.

**Comment by hppritcha** [comment]
bot:retest

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by hppritcha** [comment]
hmm.. this is weird. 
bot:retest

File: prov/gni/src/gnix_vc.c [hunk 1]
```diff
@@ -2054,20 +2053,20 @@ static int __gnix_ep_rdm_get_vc(struct gnix_fid_ep *ep, fi_addr_t dest_addr,
 		}
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c:2053]: The (unchanged) error path above this line doesn't release the lock.

File: prov/gni/src/gnix_vc.c [hunk 2]
```diff
@@ -2054,20 +2053,20 @@ static int __gnix_ep_rdm_get_vc(struct gnix_fid_ep *ep, fi_addr_t dest_addr,
...
 		if (likely(ret == FI_SUCCESS)) {
 			vc = vc_tmp;
 			vc->modes |= GNIX_VC_MODE_IN_HT;
+			fastlock_release(&ep->vc_ht_lock);
 			ret = _gnix_vc_connect(vc);
 			if (ret != FI_SUCCESS) {
 				GNIX_WARN(FI_LOG_EP_DATA,
 					"_gnix_vc_connect returned %s\n",
 					   fi_strerror(-ret));
+				fastlock_release(&ep->vc_ht_lock);
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c]: Why is this release necessary?  Did you just release it above?

File: prov/gni/src/gnix_vc.c [hunk 3]
```diff
@@ -2079,9 +2078,10 @@ static int __gnix_ep_rdm_get_vc(struct gnix_fid_ep *ep, fi_addr_t dest_addr,
 				   fi_strerror(-ret));
 			goto err;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c]: This (unchanged) error path also does not release the lock.

File: prov/gni/src/gnix_vc.c [hunk 4]
```diff
@@ -2079,9 +2078,10 @@ static int __gnix_ep_rdm_get_vc(struct gnix_fid_ep *ep, fi_addr_t dest_addr,
...
 		}
-	}
+	} else
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c]: Can you use braces around this single statement?  The new linux style checker should allow it.

**Comment by sungeunchoi** [comment]
For consistency, should `__gnix_ep_rdm_get_vc` be renamed to `__gnix_vc_ep_rdm_get_vc`?

File: prov/gni/src/gnix_vc.c (hunk 2 above)
Comment (hppritcha) [inline prov/gni/src/gnix_vc.c]: duh, you're right.

**Comment by hppritcha** [comment]
I think I'd like to change the function name in a separate pr.

**Comment by hppritcha** [comment]
@sungeunchoi take a look now

**Comment by sungeunchoi** [comment]
:+1:

**Pull request merged**